### DISCUSSION TITLE
Why 5 transactions when deploying Reach Application on Algorand OR Can I group them as one transaction ?

### DISCUSSION THREAD

**Discussion by asolpshinning** [reach-sh/reach-lang#1061] [Q&A] [answered]
Can someone please kindly explain why I need to sign 5 transactions when I am deploying my Reach application on Algorand ?
Is there any way to combine these signatures into 1 or 2 ?
Here are the prompts I got:
1. Create Application
2. Opt in Application
3. Transfer Algo (0.1 Algo)
4. Application Call
5. Application Call

And below is the rsh code:

`'reach 0.1';`

`const Params = Object({
    zeroAddress: Address,
});`

`export const main = Reach.App(() => {
    setOptions({ untrustworthyMaps: true });`

    const Creator = Participant('Creator', {
        getParams: Fun([], Params),
        iDeployed: Fun([Bytes(15)], Null),
    })`

    const User = API({
        // transfer(to, amt)
        manCtcId: Fun([Address], Bool),

        // transferFrom(from, to, amt)
        btCtcId: Fun([Address], Bool),

        // approve(approved, amt)
        stCtcId: Fun([Address], Bool),

    });

    const Views = View({
        vManCtc: Fun([Address], Address),
        vBtCtc: Fun([Address], Address),
        vStCtc: Fun([Address], Address),
    });

    init();

    Creator.only(() => {
        const { zeroAddress } = declassify(interact.getParams());
    });
    Creator.publish(zeroAddress);

    Creator.interact.iDeployed('I have deployed')

    commit();

    Creator.publish();
    //
    const end = UInt.max;
    // Unwrap bool
    //const unwbool = (m) => fromSome(m, false);
    // Unwrap Uint
    const unwInt = (m) => fromSome(m, 0);
    // Unwrap Obj
    const unwObj = (m) => fromSome(m, { spender: Creator, amt: 0 });
    // Unwrap addr
    const unwAddr = (m) => fromSome(m, zeroAddress)

    const manCtc = new Map(Address);
    const btCtc = new Map(Address);
    const stCtc = new Map(Address);

    const [manCtcNum, stCtcNum, btCtcNum] = parallelReduce([0, 0, 0])
        .invariant(balance() == 0)
        .while(true)
        .define(() => {
            Views.vManCtc.set((addr) => {
                const id = unwAddr(manCtc[addr])
                return id;
            });
            Views.vBtCtc.set((addr) => {
                const id = unwAddr(btCtc[addr])
                return id;
            });
            Views.vStCtc.set((addr) => {
                const id = unwAddr(stCtc[addr])
                return id;
            });
        })
        .api(
            User.manCtcId,
            // Assumes
            ((addr) => {
                assume(typeOf(addr) == Address);
            }),
            // Payments
            ((_) => 0),
            // Consensus
            ((addr, res) => {
                require(typeOf(addr) == Address);
                manCtc[this] = addr;
                res(true);
                return [manCtcNum + 1, stCtcNum, btCtcNum];
            })
        )
        .api(
            User.btCtcId,
            // Assumes
            ((addr) => {
                assume(typeOf(addr) == Address);
            }),
            // Payments
            ((_) => 0),
            // Consensus
            ((addr, res) => {
                require(typeOf(addr) == Address);
                btCtc[this] = addr;
                res(true);
                return [manCtcNum, stCtcNum, btCtcNum + 1];
            })
        )
        .api(
            User.stCtcId,
            // Assumes
            ((addr) => {
                assume(typeOf(addr) == Address);
            }),
            // Payments
            ((_) => 0),
            // Consensus
            ((addr, res) => {
                require(typeOf(addr) == Address);
                stCtc[this] = addr;
                res(true);
                return [manCtcNum, stCtcNum + 1, btCtcNum];
            })
        )
        .timeout(absoluteTime(end), () => {
            Anybody.publish();
            return [manCtcNum, stCtcNum, btCtcNum];
        });

    commit();
    exit();
});`

**Comment by jeapostrophe** [comment, marked as answer]
You write that you see these five signings:
1. Create Application
2. Opt in Application
3. Transfer Algo (0.1 Algo)
4. Application Call
5. Application Call

I believe that you should actually be seeing these four signings:
1. Create Application
2. Opt in Application
3. Transfer Algo (0.1 Algo) & Application Call
4. Application Call

You can remove the 4th signing by changing your code to remove the empty `publish`, which doesn't do anything.

The separation between the 1st and 3rd signings is necessary because on Algorand, you need to fund an account (the application's) before it can do anything, but we don't know what the account is going to be until after it is created.

It is theoretically possible for us to combine the 2nd and 3rd on your behalf by improving the compiler, but we don't know, because it is a little complex to reason about some of the corner cases of when it is okay to opt-in at the same time as doing work.

**Reply by asolpshinning** [reply]
Thanks for providing answer Jay. I should remove the empty publish.
